performance/class-media-library-caching.php

```diff
@@ -0,0 +1,163 @@
+			$use_defaults     = $attachment_count > $max_posts_to_query;
+
+			if ( $use_defaults ) {
+				// If there are too many posts to query, use the default MIME types.
+				$mime_types = self::get_default_mime_types();
+			} else {
+				// Otherwise, use the same query from core.
+				$mime_types = $wpdb->get_col( $wpdb->prepare( "SELECT DISTINCT post_mime_type FROM $wpdb->posts WHERE post_type = %s", $type ) );
+			}
+
+			// Cache the results.
+			wp_cache_set( self::AVAILABLE_MIME_TYPES_CACHE_KEY, $mime_types, self::CACHE_GROUP );
+			wp_cache_set( self::USING_DEFAULT_MIME_TYPES_CACHE_KEY, $use_defaults, self::CACHE_GROUP );
+		}
+
+		// If there were any previous MIME types, merge them with the cached MIME types.
+		if ( is_array( $filtered_mime_types ) ) {
+			$mime_types = array_unique( array_merge( $filtered_mime_types, $mime_types ) );
+		}
+
+		return $mime_types;
+	}
+
+	/**
+	 * Get the default MIME types.
+	 *
+	 * @return string[] An array of default MIME types.
+	 */
+	public static function get_default_mime_types() {
+		// Massage the results from get_post_mime_types() into a flat array.
+		return array_reduce(
+			array_keys( get_post_mime_types() ),
+			function ( $carry, $mime_type ) {
+				return array_merge( $carry, explode( ',', $mime_type ) );
+			},
+			array()
+		);
+	}
+
+	/**
+	 * Update the MIME types cache when a new post is added.
+	 *
+	 * @param int $post_id The post ID.
+	 */
+	public static function update_post_mime_types_cache_on_add( $post_id ) {
+		if ( wp_cache_get( self::USING_DEFAULT_MIME_TYPES_CACHE_KEY, self::CACHE_GROUP ) ) {
+			return;
+		}
+
+		$mime_type = get_post_mime_type( $post_id );
+
+		if ( false !== $mime_type ) {
+			$mime_types = wp_cache_get( self::AVAILABLE_MIME_TYPES_CACHE_KEY, self::CACHE_GROUP );
+
+			if ( false !== $mime_types ) {
+				// Add the new MIME type to the cache if not present.
+				if ( ! in_array( $mime_type, $mime_types, true ) ) {
+					$mime_types[] = $mime_type;
+					wp_cache_set( self::AVAILABLE_MIME_TYPES_CACHE_KEY, $mime_types, self::CACHE_GROUP );
+				}
+			}
+		}
+	}
+
+	/**
+	 * Update the MIME types cache when a post is edited.
+	 *
+	 * @param int     $post_id     The post ID.
+	 * @param WP_Post $post_after  The post object after the update.
+	 * @param WP_Post $post_before The post object before the update.
+	 */
+	public static function update_post_mime_types_cache_on_edit( $post_id, $post_after, $post_before ) {
+		// Only if the MIME type changed.
+		if ( $post_before->post_mime_type !== $post_after->post_mime_type ) {
+			wp_cache_delete( self::AVAILABLE_MIME_TYPES_CACHE_KEY, self::CACHE_GROUP );
+		}
+	}
+
+	/**
+	 * Update the MIME types cache when a post is deleted.
+	 */
+	public static function update_post_mime_types_cache_on_delete() {
+		wp_cache_delete( self::AVAILABLE_MIME_TYPES_CACHE_KEY, self::CACHE_GROUP );
```

Comment:
I feel like this might lead to a bit more invalidation than really needed.

Maybe we could do a follow-up query and see if we can find just 1 attachment with the mime type that is being deleted. And if so, there's no need to invalidate. That may be overkill though.

Also, if `USING_DEFAULT_MIME_TYPES_CACHE_KEY` is true, we shouldn't invalidate right?

Reply:
@WPprodigy That was precisely my initial approach to this, but I refactored it after the PR feedback, so it wasn't overkill and we avoid the extra query 😄 

You can see the previous state here:
https://github.com/Automattic/vip-go-mu-plugins/blob/95eaa57dca82cd1e18e49013f1bcbc14ad002807/performance/class-media-library-caching.php#L203-L212

Reply:
> Also, if USING_DEFAULT_MIME_TYPES_CACHE_KEY is true, we shouldn't invalidate right?

That's a tricky question because there's the edge case where adding/removing attachments might go over or below the threshold for executing the query or going with the defaults.

So, for example, if the `MAX_POSTS_TO_QUERY_DEFAULT` variable is set to 500K, and you currently have 500,001 attachments but delete one (or a bunch of them), then in theory, you should stop showing the defaults and do the DB query, but that wouldn't happen if you don't invalidate the cache so that the threshold could be rechecked.

Reply:
I guess it depends on how expensive the `SELECT COUNT(1) FROM {$wpdb->posts} WHERE post_type = %s` query could be, as invalidating here causes that to run again at a minimum.

If that gets pricey, then I'd say it's worth just aborting if already using default mime types. I'm not really worried about being EXACT on the `max_posts_to_query`. But if that stays cheap enough, then invalidating here is fine :)

Reply:
Based on the tests I performed while checking [this](https://github.com/Automattic/vip-go-mu-plugins/pull/4953#discussion_r1377025323), the count attachment select was relatively slow on very large tables, but still only took about 1/20th of the time that the distinct query takes, as it will only read the index. In a scenario with about 10M attachments, it took less than 10 seconds, while the distinct query could take up to minutes.

Regarding the previous delete check that I referenced above, that one is better than running a new distinct query but still slower than the attachment count check, as it needs to read the table.

Having said that, I think I'll avoid invalidating the cache if using defaults then.

Reply:
A check was added to the update/delete filters to check if it's using defaults.